packages/snaps-rpc-methods/src/permitted/setState.ts

```diff
@@ -0,0 +1,252 @@
+
+  if (!hasPermission(manageStateBuilder.targetName)) {
+    return end(providerErrors.unauthorized());
+  }
+
+  try {
+    const validatedParams = getValidatedParams(params);
+    const { key, value, encrypted = true } = validatedParams;
+
+    if (encrypted) {
+      await getUnlockPromise(true);
+    }
+
+    if (key === undefined && !isPlainObject(value)) {
+      return end(
+        rpcErrors.invalidParams(
+          'Invalid params: Value must be an object if key is not provided.',
+        ),
+      );
+    }
+
+    const newState = await getNewState(key, value, getSnapState);
+    await updateSnapState(newState, encrypted);
+    response.result = null;
+  } catch (error) {
+    return end(error);
+  }
+
+  return end();
+}
+
+/**
+ * Validate the parameters of the `snap_setState` method.
+ *
+ * @param params - The parameters to validate.
+ * @returns The validated parameters.
+ */
+function getValidatedParams(params?: unknown) {
+  try {
+    return create(params, SetStateParametersStruct);
+  } catch (error) {
+    if (error instanceof StructError) {
+      throw rpcErrors.invalidParams({
+        message: `Invalid params: ${error.message}.`,
+      });
+    }
+
+    /* istanbul ignore next */
+    throw rpcErrors.internal();
+  }
+}
+
+/**
+ * Get the new state of the Snap.
+ *
+ * If the key is `undefined`, the value is expected to be an object. In this
+ * case, the value is returned as the new state.
+ *
+ * If the key is not `undefined`, the value is set in the state at the key. If
+ * the key does not exist, it is created (and any missing intermediate keys are
+ * created as well).
+ *
+ * @param key - The key to set.
+ * @param value - The value to set the key to.
+ * @param getSnapState - The `getSnapState` hook.
+ * @returns The new state of the Snap.
+ */
+async function getNewState(
+  key: string | undefined,
+  value: Json,
+  getSnapState: SetStateHooks['getSnapState'],
+) {
+  if (key === undefined) {
+    assert(isPlainObject(value));
+    return value;
+  }
+
+  const state = await getSnapState(false);
+  return set(state, key, value);
+}
+
+/**
+ * Set the value of a key in an object. The key may contain Lodash-style path
+ * syntax, e.g., `a.b.c` (with the exception of array syntax). If the key does
+ * not exist, it is created (and any missing intermediate keys are created as
+ * well).
+ *
+ * This is a simplified version of Lodash's `set` function, but Lodash doesn't
+ * seem to be maintained anymore, so we're using our own implementation.
+ *
+ * @param object - The object to get the key from.
+ * @param key - The key to set.
+ * @param value - The value to set the key to.
+ * @returns The new object with the key set to the value.
+ */
+export function set(
+  // eslint-disable-next-line @typescript-eslint/default-param-last
+  object: Record<string, Json> | null,
+  key: string | undefined,
+  value: Json,
+): JsonObject {
+  if (key === undefined) {
+    assert(isPlainObject(value));
```

Comment:
Why are we asserting this? Wondering about performance implications

Reply:
Type inference 😅

Reply:
Both `assert` and `isObject` are tiny functions, so this should be extremely quick. We know this throws in the current implementation, but that may not hold in future refactors.